// reakit/es/Card.js
import { theme, withProp } from 'styled-tools';
import styled from './styled.js';
import use from './use.js';
import Box from './Box.js';
import { oneOfType, number, string, bool, oneOf, func, object, arrayOf, node } from 'prop-types';
import s from 'styled-selector';
import { a as numberToPx } from './chunk-a7ce379e.js';
import 'styled-components';
import 'reuse';
import '@emotion/is-prop-valid';
import './chunk-fb907809.js';
import 'react';
import './chunk-bbbc8dce.js';

var CardFit = styled(Box).withConfig({
  displayName: "CardFit",
  componentId: "sc-17ckqrf-0"
})(["", ";"], theme("CardFit"));
var CardFit$1 = use(CardFit, "div");

var Card = styled(Box).withConfig({
  displayName: "Card",
  componentId: "sc-19jv5xv-0"
})(["display:inline-block;&& > *:not(", "){margin:", ";}", ";"], s(CardFit$1), withProp("gutter", numberToPx), theme("Card")); // @ts-ignore

Card.propTypes = {
  gutter: oneOfType([number, string])
};
Card.defaultProps = {
  gutter: "1rem",
  opaque: true,
  palette: "background",
  tone: -1
};
var Card$1 = use(Card, "div");

var index = Object.assign(Card$1, {
  Fit: CardFit$1
});

export default index;
