// reakit/es/chunk-52a78301.js
import { a as _extends, f as _objectSpread } from './chunk-fb907809.js';
import { forwardRef, createElement, Component, createRef } from 'react';
import { oneOfType, number, string, bool, oneOf, func, object, arrayOf, node } from 'prop-types';
import { Container } from 'constate';

var initialState = {
  visible: false
};
var actions = {
  show: function show() {
    return function () {
      return {
        visible: true
      };
    };
  },
  hide: function hide() {
    return function () {
      return {
        visible: false
      };
    };
  },
  toggle: function toggle() {
    return function (state) {
      return {
        visible: !state.visible
      };
    };
  }
};

var HiddenContainer = function HiddenContainer(props) {
  return createElement(Container, _extends({}, props, {
    initialState: _objectSpread({}, initialState, props.initialState),
    actions: actions
  }));
}; // @ts-ignore


HiddenContainer.propTypes = {
  initialState: object
};
var HiddenContainer$1 = Object.assign(HiddenContainer, {
  initialState: initialState,
  actions: actions
});

export { HiddenContainer$1 as a };
