// reakit/es/Code.js
import { a as _extends } from './chunk-fb907809.js';
import { forwardRef, createElement, Component, createRef } from 'react';
import { oneOfType, number, string, bool, oneOf, func, object, arrayOf, node } from 'prop-types';
import { theme } from 'styled-tools';
import styled from './styled.js';
import use from './use.js';
import Box from './Box.js';
import 'styled-components';
import 'reuse';
import '@emotion/is-prop-valid';
import './chunk-bbbc8dce.js';
import './chunk-a7ce379e.js';

var CodeComponent = function CodeComponent(props) {
  if (props.block) {
    return createElement(Box, _extends({
      use: "pre"
    }, props), createElement("code", {
      className: props.codeClassName
    }, props.children));
  }

  var className = props.codeClassName ? (props.className + " " + props.codeClassName).trim() : ("" + props.className).trim();
  return createElement(Box, _extends({
    use: "code"
  }, props, {
    className: className
  }));
};

var Code = styled(Box).withConfig({
  displayName: "Code",
  componentId: "e6efej-0"
})(["", ";"], theme("Code")); // @ts-ignore

Code.propTypes = {
  block: bool,
  codeClassName: string
};
Code.defaultProps = {
  use: CodeComponent,
  opaque: true,
  palette: "background",
  tone: -2
};
var Code$1 = use(Code);

export default Code$1;
