// reakit/es/Sidebar.js
import { a as _extends } from './chunk-fb907809.js';
import { forwardRef, createElement, Component, createRef } from 'react';
import { oneOfType, number, string, bool, oneOf, func, object, arrayOf, node } from 'prop-types';
import { theme, ifProp } from 'styled-tools';
import { a as hoist } from './chunk-9e6fe658.js';
import styled from './styled.js';
import use from './use.js';
import Overlay from './Overlay.js';
import { a as OverlayContainer, b as OverlayToggle, c as OverlayShow, d as OverlayHide } from './chunk-fecc90d4.js';
import 'styled-components';
import 'reuse';
import './Hidden.js';
import 'react-dom';
import './chunk-bbbc8dce.js';
import './chunk-a7ce379e.js';
import './chunk-d8771a86.js';
import './Box.js';
import '@emotion/is-prop-valid';
import './chunk-52a78301.js';
import 'constate';
import './chunk-9c9a5f71.js';

var SidebarComponent = function SidebarComponent(props) {
  return createElement(Overlay, _extends({
    defaultSlide: props.align === "right" ? "left" : "right"
  }, props));
};

var Sidebar = styled(hoist(SidebarComponent, Overlay)).withConfig({
  displayName: "Sidebar",
  componentId: "sc-1ujdgs6-0"
})(["top:0;height:100vh;transform:none;overflow:auto;left:", ";right:", ";", ";"], ifProp({
  align: "right"
}, "auto", 0), ifProp({
  align: "right"
}, 0, "auto"), theme("Sidebar")); // @ts-ignore

Sidebar.propTypes = {
  align: oneOf(["left", "right"])
};
Sidebar.defaultProps = {
  align: "left",
  translateX: 0,
  translateY: 0
};
var Sidebar$1 = use(Sidebar, "div");

var SidebarContainer = OverlayContainer;

var SidebarToggle = styled(OverlayToggle).withConfig({
  displayName: "SidebarToggle",
  componentId: "sc-1ldhjpz-0"
})(["", ";"], theme("SidebarToggle"));
var SidebarToggle$1 = use(SidebarToggle, "button");

var SidebarShow = styled(OverlayShow).withConfig({
  displayName: "SidebarShow",
  componentId: "sc-1jmdl92-0"
})(["", ";"], theme("SidebarShow"));
var SidebarShow$1 = use(SidebarShow, "button");

var SidebarHide = styled(OverlayHide).withConfig({
  displayName: "SidebarHide",
  componentId: "sc-1wp4kne-0"
})(["", ";"], theme("SidebarHide"));
var SidebarHide$1 = use(SidebarHide, "button");

var index = Object.assign(Sidebar$1, {
  Container: SidebarContainer,
  Toggle: SidebarToggle$1,
  Show: SidebarShow$1,
  Hide: SidebarHide$1
});

export default index;
