// reakit/lib/Backdrop.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var styledTools = require('styled-tools');
var styled = require('./styled.js');
var use = require('./use.js');
var Hidden = require('./Hidden.js');
require('styled-components');
require('reuse');
require('./chunk-582bef72.js');
require('react');
require('react-dom');
require('prop-types');
require('./chunk-b4ae9232.js');
require('./chunk-f0c301b2.js');
require('./chunk-5814b5fa.js');
require('./chunk-25591e69.js');
require('./Box.js');
require('@emotion/is-prop-valid');
require('./chunk-9ddd950d.js');
require('constate');
require('./chunk-2aaf90c8.js');

var Backdrop = styled.default(Hidden.default).withConfig({
  displayName: "Backdrop",
  componentId: "sc-1mc2k5c-0"
})(["position:fixed;top:0;left:0;width:100%;height:100%;z-index:998;-webkit-tap-highlight-color:rgba(0,0,0,0);-moz-tap-highlight-color:rgba(0,0,0,0);", ";"], styledTools.theme("Backdrop"));
Backdrop.defaultProps = {
  role: "button",
  tabIndex: -1,
  opaque: true,
  palette: "shadow",
  tone: 2
};
var Backdrop$1 = use.default(Backdrop, "div");

exports.default = Backdrop$1;
