// reakit/lib/Card.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var styledTools = require('styled-tools');
var styled = require('./styled.js');
var use = require('./use.js');
var Box = require('./Box.js');
var PropTypes = require('prop-types');
var s = _interopDefault(require('styled-selector'));
var __chunk_2 = require('./chunk-f0c301b2.js');
require('styled-components');
require('reuse');
require('@emotion/is-prop-valid');
require('./chunk-582bef72.js');
require('react');
require('./chunk-b4ae9232.js');

var CardFit = styled.default(Box.default).withConfig({
  displayName: "CardFit",
  componentId: "sc-17ckqrf-0"
})(["", ";"], styledTools.theme("CardFit"));
var CardFit$1 = use.default(CardFit, "div");

var Card = styled.default(Box.default).withConfig({
  displayName: "Card",
  componentId: "sc-19jv5xv-0"
})(["display:inline-block;&& > *:not(", "){margin:", ";}", ";"], s(CardFit$1), styledTools.withProp("gutter", __chunk_2.numberToPx), styledTools.theme("Card")); // @ts-ignore

Card.propTypes = {
  gutter: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};
Card.defaultProps = {
  gutter: "1rem",
  opaque: true,
  palette: "background",
  tone: -1
};
var Card$1 = use.default(Card, "div");

var index = Object.assign(Card$1, {
  Fit: CardFit$1
});

exports.default = index;
