// reakit/lib/chunk-aa4a91c3.js
'use strict';

var __chunk_6 = require('./chunk-9ddd950d.js');
var styledTools = require('styled-tools');
var styled = require('./styled.js');
var use = require('./use.js');
var __chunk_7 = require('./chunk-2aaf90c8.js');

var OverlayContainer = __chunk_6.HiddenContainer;

var OverlayToggle = styled.default(__chunk_7.HiddenToggle).withConfig({
  displayName: "OverlayToggle",
  componentId: "sc-1u1lo2-0"
})(["", ";"], styledTools.theme("OverlayToggle"));
var OverlayToggle$1 = use.default(OverlayToggle, "button");

var OverlayShow = styled.default(__chunk_7.HiddenShow).withConfig({
  displayName: "OverlayShow",
  componentId: "c48b0n-0"
})(["", ";"], styledTools.theme("OverlayShow"));
var OverlayShow$1 = use.default(OverlayShow, "button");

var OverlayHide = styled.default(__chunk_7.HiddenHide).withConfig({
  displayName: "OverlayHide",
  componentId: "sc-1d97up5-0"
})(["", ";"], styledTools.theme("OverlayHide"));
var OverlayHide$1 = use.default(OverlayHide, "button");

exports.OverlayContainer = OverlayContainer;
exports.OverlayToggle = OverlayToggle$1;
exports.OverlayShow = OverlayShow$1;
exports.OverlayHide = OverlayHide$1;
