// reakit/lib/chunk-9ddd950d.js
'use strict';

var __chunk_1 = require('./chunk-582bef72.js');
var React = require('react');
var PropTypes = require('prop-types');
var constate = require('constate');

var initialState = {
  visible: false
};
var actions = {
  show: function show() {
    return function () {
      return {
        visible: true
      };
    };
  },
  hide: function hide() {
    return function () {
      return {
        visible: false
      };
    };
  },
  toggle: function toggle() {
    return function (state) {
      return {
        visible: !state.visible
      };
    };
  }
};

var HiddenContainer = function HiddenContainer(props) {
  return React.createElement(constate.Container, __chunk_1._extends({}, props, {
    initialState: __chunk_1._objectSpread({}, initialState, props.initialState),
    actions: actions
  }));
}; // @ts-ignore


HiddenContainer.propTypes = {
  initialState: PropTypes.object
};
var HiddenContainer$1 = Object.assign(HiddenContainer, {
  initialState: initialState,
  actions: actions
});

exports.HiddenContainer = HiddenContainer$1;
