// reakit/lib/Overlay.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var styledTools = require('styled-tools');
var styled = require('./styled.js');
var use = require('./use.js');
var Hidden = require('./Hidden.js');
var __chunk_8 = require('./chunk-aa4a91c3.js');
require('styled-components');
require('reuse');
require('./chunk-582bef72.js');
require('react');
require('react-dom');
require('prop-types');
require('./chunk-b4ae9232.js');
require('./chunk-f0c301b2.js');
require('./chunk-5814b5fa.js');
require('./chunk-25591e69.js');
require('./Box.js');
require('@emotion/is-prop-valid');
require('./chunk-9ddd950d.js');
require('constate');
require('./chunk-2aaf90c8.js');

var Overlay = styled.default(Hidden.default).withConfig({
  displayName: "Overlay",
  componentId: "tgx0ep-0"
})(["position:fixed;z-index:19900410;left:50%;top:50%;", ";"], styledTools.theme("Overlay"));
Overlay.defaultProps = {
  role: "dialog",
  "aria-modal": true,
  hideOnEsc: true,
  translateX: "-50%",
  translateY: "-50%",
  defaultSlide: "top",
  opaque: true,
  palette: "background",
  tone: -1
};
var Overlay$1 = use.default(Overlay, "div");

var Overlay$2 = Object.assign(Overlay$1, {
  Container: __chunk_8.OverlayContainer,
  Toggle: __chunk_8.OverlayToggle,
  Show: __chunk_8.OverlayShow,
  Hide: __chunk_8.OverlayHide
});

exports.default = Overlay$2;
