// reakit/lib/Portal.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var __chunk_1 = require('./chunk-582bef72.js');
var React = require('react');
var ReactDOM = require('react-dom');
var styledTools = require('styled-tools');
var __chunk_4 = require('./chunk-5814b5fa.js');
var styled = require('./styled.js');
var use = require('./use.js');
var Box = require('./Box.js');
require('styled-components');
require('reuse');
require('@emotion/is-prop-valid');
require('prop-types');
require('./chunk-b4ae9232.js');
require('./chunk-f0c301b2.js');

var PortalComponent =
/*#__PURE__*/
function (_React$Component) {
  __chunk_1._inheritsLoose(PortalComponent, _React$Component);

  function PortalComponent() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

    __chunk_1._defineProperty(__chunk_1._assertThisInitialized(__chunk_1._assertThisInitialized(_this)), "state", {
      wrapper: null
    });

    return _this;
  }

  var _proto = PortalComponent.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var wrapper = document.createElement("div");
    document.body.appendChild(wrapper);
    this.setState({
      wrapper: wrapper
    });
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    var wrapper = this.state.wrapper;

    if (wrapper) {
      document.body.removeChild(wrapper);
    }
  };

  _proto.render = function render() {
    var wrapper = this.state.wrapper;

    if (wrapper) {
      return ReactDOM.createPortal(React.createElement(Box.default, this.props), wrapper);
    }

    return null;
  };

  return PortalComponent;
}(React.Component);

var Portal = styled.default(__chunk_4.hoist(PortalComponent, Box.default)).withConfig({
  displayName: "Portal",
  componentId: "sc-1anvbga-0"
})(["", ";"], styledTools.theme("Portal"));
var Portal$1 = use.default(Portal, "div");

exports.default = Portal$1;
